Source/WebCore/automation/AutomationInstrumentation.h

```diff
@@ -0,0 +1,68 @@
+/*
+* Copyright (C) 2024 Igalia S.L.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are
+* met:
+*
+*     * Redistributions of source code must retain the above copyright
+* notice, this list of conditions and the following disclaimer.
+*     * Redistributions in binary form must reproduce the above
+* copyright notice, this list of conditions and the following disclaimer
+* in the documentation and/or other materials provided with the
+* distribution.
+*     * Neither the name of Google Inc. nor the names of its
+* contributors may be used to endorse or promote products derived from
+* this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+* A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#pragma once
+
+#if ENABLE(WEBDRIVER_BIDI)
+
+#include <JavaScriptCore/ConsoleMessage.h>
+#include <wtf/CompletionHandler.h>
+#include <wtf/Function.h>
+#include <wtf/MemoryPressureHandler.h>
+#include <wtf/ObjectIdentifier.h>
+#include <wtf/RefPtr.h>
+
+namespace Inspector {
+class ConsoleMessage;
+}
+
+namespace WebCore {
+
+
+using AutomationInstrumentationConsoleMessageObserver = WTF::Observer<void(const Inspector::ConsoleMessage&)>;
+
+class WEBCORE_EXPORT AutomationInstrumentation {
+public:
+    static void setConsoleMessageObserver(const AutomationInstrumentationConsoleMessageObserver&);
+    static void clearConsoleMessageObserver();
+
+    static void addMessageToConsole(const std::unique_ptr<Inspector::ConsoleMessage>&);
+    static void addMessageToConsoleImpl(const std::unique_ptr<Inspector::ConsoleMessage>&);
+};
+
+inline void AutomationInstrumentation::addMessageToConsole(const std::unique_ptr<Inspector::ConsoleMessage>& message)
+{
+    addMessageToConsoleImpl(message);
```

Comment:
im not really sure what benefit this extra level of indirection provides

for `InspectorInstrumentation` we will quickly return if there are no frontends present.  perhaps we should do the same if there are no automation sessions?

Reply:
I replaced this indirection with a direct implementation, given it's a "shallow" method.